FogCreek/Glitch-Community
# Zest ostrich: Feature a Project Embed in Collections

## Links
* https://zest-ostrich.glitch.me/@sarahzinger/blossom-mix
* https://glitch.manuscript.com/f/cases/3328003/
* https://www.notion.so/glitch/Narrative-Collections-162c85d294a0453b9d57f7f8cdaa6783

## GIF/Screenshots:
<img width="1188" alt="Screen Shot 2019-04-09 at 4 20 44 PM" src="https://user-images.githubusercontent.com/6620164/55832518-c0c93a80-5ae3-11e9-8f58-c31e36a5fd9a.png">
<img width="1179" alt="Screen Shot 2019-04-09 at 4 20 57 PM" src="https://user-images.githubusercontent.com/6620164/55832519-c0c93a80-5ae3-11e9-8a4b-e27ffc793316.png">
<img width="1192" alt="Screen Shot 2019-04-09 at 4 21 09 PM" src="https://user-images.githubusercontent.com/6620164/55832520-c0c93a80-5ae3-11e9-9de7-eaba8828a233.png">
<img width="1171" alt="Screen Shot 2019-04-09 at 4 21 27 PM" src="https://user-images.githubusercontent.com/6620164/55832521-c0c93a80-5ae3-11e9-93fa-42975e02f3af.png">
<img width="1149" alt="Screen Shot 2019-04-09 at 4 21 46 PM" src="https://user-images.githubusercontent.com/6620164/55832522-c0c93a80-5ae3-11e9-9785-1689ae8cd1bf.png">

## Changes:
* Adds a `Feature` button to each project item on a collections page
* Adds `FeaturedProject` to the Collections Page that will render if there's a featured project
* Adds an optional `Note` to the `FeaturedProject` component
* Adds an `Add Note` button to the `FeaturedProject` component
* Some syntax changes to `FeaturedProject` in how we pass components to `ProjectEmbed`
* Change how we update state when an individual project changes in a collection so it's a bit more DRY
* Remove an unused class
* small DRY change to the project-options-pop
* update `Note` to be more explicit when users are authorized to update a note or not rather than depend on passing an update function, along with some var name changes
* Fix feature/unfeature animation bug

## How To Test:
* here's a collection with a featured project and some notes: https://zest-ostrich.glitch.me/@sarahzinger/blossom-mix
* on your own collection try to feature and unfeature a project, add and remove a note, etc
* double check featured projects on team and user pages still look good. 

## Feedback I'm looking for:
* Thoughts on weird things I might be doing codewise? In particular how do we feel about `FeaturedProject` it continues to feel weird to me the way I'm passing child components and I wonder if I'm approaching this the wrong way or if I'm over thinking it. 
* ~~I feel like the manipulation of `collection` state in `CollectionEditor` (particularly for `collection.projects` is getting to be a real pain in the neck. I wonder if it's time to reconsider the approach we've taken. I feel like in an ideal world we make a data request to a backend and then we just setState with the data we get back from the Patch, but the data we get back is incomplete. I can think of a few possible approaches to handling this, none of which feel easy or obviously right to me. Eager to hear other thoughts on this! Here are some of my thoughts on paths we could take:`~~
    * ~~maybe this is a sign we need redux here? Certain parts of state know to subscribe to certain patch requests and figure out how to update accordingly? This isn't fully fleshed out just an impulse? @modernserf I bet you have thoughts!~~
    * ~~we recall the `loadCollection` call from `collection.js` after each patch to update a collection. This would make more http requests but ensures we set state in the same way in the same place. We refetch all the users for each project which is maybe not performant but ensures a single source of truth, our backend. I wonder if this approach is easier/necessary for the work @keithk is doing for reordering projects in a collection? I suspect whatever your state needs for reordering, keith, are more pressing, and I'm happy to just follow whatever lead you take with that :) This PR can be merged after all that stuff is done if you want, happy to deal with the inevitable weird merge requests we get (sorry didn't think this would be so intertwined when I picked up the ticket...but I'm thinking it might be, or maybe not who knows)~~
    * ~~we continue down the path we're already on, as demonstrated in this pr, by making data requests when we need and (mostly) ignoring those responses. We manually manage state updates ourselves as needed.~~
    * ~~some other approach I'm just not thinking about?~~
* ui feedback. For example, I wonder if the note looks a little strange? Particularly in edit mode it expands, this looked ok-ish on a project item but is starting to look a bit weird to me above the featured project. Should it be full width I wonder? Probably at least on mobile/small screens? @ttseng (and others!) lend me your wisdom :) 

## Things left to do before deploying:
- [x] double check backend calls are working once those are finished
- [x] figure out why animation for unfeature is...not working I think? I think this might have been a regression I caused when componentizing featuredProjectEmbed
- [x] remove add note button when note is already present from featured project
- [x] fix featuring private projects bug
- [x] talk out if my approach to state management here is yucky and if it is revise it so it's better. (now that I'm not manually storing a featuredProject but just filtering based on the featuredProjectId, the state management feels much more reasonable and less weird)
- [x] clean up css and qa it
- [x] cross browser test
- [x] a quick voice over check to make sure I can navigate to all the buttons and stuff. 
- [x] fix bug where notes are out of place when you feature an item when there is already a feature
- [x] fix bug where logged in non-authors see feature button on user page

Post deploy checks
- [ ] Featured project: 
    - [ ] User (owns)
        - [ ] Can’t add note
        - [ ] Can’t feature private projects
        - [ ] Can pin private projects
        - [ ] Can feature project
        - [ ] Can unfeature project
    - [ ] Team (owns)
        - [ ] Can’t add note
        - [ ] Can’t feature private projects
        - [ ] Can pin private projects
        - [ ] Can feature project
        - [ ] Can unfeature project
    - [ ] Collection (owns)
        - [ ] Can add note 
        - [ ] Can delete note
        - [ ] Can’t feature private projects
        - [ ] Can feature regular projects
        - [ ] Can un feature project
        - [ ] Can add project
        - [ ] Can remove project
    - [ ]  Not logged in can view all that ^
    - [ ] Logged in but not owner can view and add project to collections
- [ ] ProjectEmbed
    - [ ] looks good on project page
- [ ] Animations:
    - [ ] Featuring/unfeaturing projects
    - [ ] adding/Removing pin
    - [ ] Showing/hiding all deleted projects
    - [ ] Delete a collection
    - [ ] Delete a project
